tests/migration_tests/test_migration.py

```diff
@@ -118,3 +125,175 @@ def test_do_migrate_from_old_version_postgres(self, patched_input):
+        # Skip CSV files
+        if "csv_files" in current_path:
+            continue
+
+        for file in files:
+            full_paths[file] = os.path.join(current_path, file)
+    return full_paths
+
+
+def convert_query_result_to_dictionary(result):
+    full_paths = get_full_paths_of_sample_data()
+    # Put them in a proper dictionary, where key is an alembic version,
+    # and value is a list that contains datafile references
+    version_datafile_dict = {}
+    for version, datafile in result:
+        if version not in version_datafile_dict:
+            version_datafile_dict[version] = []
+        version_datafile_dict[version].append(full_paths.get(datafile))
+    return version_datafile_dict
+
+
+class StepByStepMigrationTestCase(unittest.TestCase):
+    """This class has two tests that create a custom table named *version_datafile*, which has
+    alembic versions with datafile references in each row. Then, it starts to upgrade the old version
+    database one by one. In each upgrade, it checks whether if there is any related value in the
+    custom database. If there is, it imports those datafiles and continues to upgrade.
+
+    Currently, old databases in the repository is generated by *Pepys 0.0.17*. Therefore, there is
+    only one data type (nisida) which isn't imported. In the future,
+    :code:`version_datafile_table.sql` files under :code:`migration_tests/postgres` and
+    :code:`migration_tests/sqlite` might be changed to include new datafile types.
+    """
+
+    def test_migrate_sqlite(self):
+        shutil.copyfile(src=SQLITE_PATH, dst=COPY_DB_PATH)
+
+        data_store = DataStore("", "", "", 0, COPY_DB_PATH, "sqlite")
+        admin_shell = AdminShell(data_store)
+        config = admin_shell.cfg
+        latest_sqlite_version = get_latest_alembic_version()
+        # Read SQL file for creating version/datafile table and inserting values
+        with open(SQLITE_SQL_PATH, "r") as f:
+            sql = f.read()
+
+        # Run SQL script, then fetch all values from the version/datafile table
+        connection = sqlite3.connect(COPY_DB_PATH)
+        connection.executescript(sql)
+        result = connection.execute(f"SELECT version, datafile FROM version_datafile ORDER BY id;")
+        result = result.fetchall()
+        version_datafile_dict = convert_query_result_to_dictionary(result)
+
+        # Migrate the database one by one, import datafiles if specified in version/datafile table
+        while True:
+            try:
+                command.upgrade(config, "+1")
+                new_version = get_alembic_version(connection)
+                if new_version in version_datafile_dict:
+                    import_files(version_datafile_dict[new_version], data_store)
+                if new_version == latest_sqlite_version:
+                    print("Upgrade to head is successful!")
+                    break
+            except Exception as e:
```

Comment:
Do we want to fail the test as well if the alembic migration gives an error? Or are we expecting some (minor) Alembic errors that we just have to ignore?

(Same applies in the Postgres test)

Reply:
Actually no, we don't want it anymore. Previously, I wasn't reading the `latest_revision.json`, so it was trying to upgrade from the latest to a non-existing version and it was giving an error. But I added another break for the while loop and I think we are okay to remove it. I'll do that for both tests.